### PR TITLE
Lesson3 review

### PR DESCRIPTION
**Pull request by podolyanbogdan**

### CODE REVIEW

File: .idea/deploymentTargetDropDown.xml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (shadwork) [inline .idea/deploymentTargetDropDown.xml:1]: all files from idea folder should be excluded from repo and added into gitignore

File: app/build.gradle [hunk 2]
```diff
@@ -31,19 +31,17 @@ android {
...
     }
 
-    viewBinding{
-        enabled true
-    }
-    dataBinding{
-        enabled true
+    buildFeatures {
+        dataBinding = true
+        viewBinding = true
```
Comment (shadwork) [inline app/build.gradle:36]: use only dataBinding

File: app/src/main/java/com/example/login/canvas/TicTacBoard.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        strokeWidth = strokeSize
+        strokeCap = Paint.Cap.ROUND
+    }
+
+    override fun onMeasure(widthSpec: Int, heightSpec: Int) {
+        super.onMeasure(widthSpec, heightSpec)
+        widthF = MeasureSpec.getSize(widthSpec).toFloat()
+        heightF = MeasureSpec.getSize(heightSpec).toFloat()
+
+        cellWidth = widthF / 3f
```
Comment (shadwork) [inline app/src/main/java/com/example/login/canvas/TicTacBoard.kt:50]: 3f should be getting through Game class that provides all definition of game include board dimension

File: app/src/main/java/com/example/login/canvas/TicTacBoard.kt [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        cellHeight = heightF / 3f
+
+        drawWinLine(mWinLineState)
+    }
+
+    override fun dispatchDraw(canvas: Canvas?) {
+        super.dispatchDraw(canvas)
+        canvas?.run {
+            for (i in 1..2) {
```
Comment (shadwork) [inline app/src/main/java/com/example/login/canvas/TicTacBoard.kt:59]: 1..2 should be declared in Game class

File: app/src/main/java/com/example/login/ui/fragments/GameFragment.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        binding.btnStartGame.setOnClickListener {
+            viewModel.onReloadClick()
+        }
+        binding.field.forEachIndexed { index, view ->
+            view.setOnClickListener { viewModel.onCellClick(index) }
```
Comment (shadwork) [inline app/src/main/java/com/example/login/ui/fragments/GameFragment.kt:32]: click should be handled in viewmodel

File: app/src/main/java/com/example/login/ui/fragments/GameViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    val currentMove: LiveData<CellState> = mCurrentMove
+    val cellStateByIndex: LiveData<Pair<Int, CellState>> = mCellStateByIndex
+    val winState: LiveData<WinLineState> = mWinState
+
+    init {
+        initGame()
+    }
+
+    private fun initGame() {
+        matrix = Array(9) { CellState.None }
```
Comment (shadwork) [inline app/src/main/java/com/example/login/ui/fragments/GameViewModel.kt:31]: size of board should not be hardcoded

File: app/src/main/res/layout/fragment_game.xml [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:background="@color/splash_color"
+    tools:context=".ui.another.MainActivity"
+    tools:ignore="ContentDescription">
+
+
+    <ImageView
+        android:id="@+id/circle"
+        android:layout_width="70dp"
```
Comment (shadwork) [inline app/src/main/res/layout/fragment_game.xml:15]: should not be hardcoded

**Review by shadwork — CHANGES_REQUESTED** [review]
Плохо. софтина крешится даже при беглом тестировании
шаблон не использован
нет рисования на канвасе в принципе
   java.lang.ArrayIndexOutOfBoundsException: length=9; index=9
        at com.example.login.ui.fragments.GameViewModel.computerMakeTurn(GameViewModel.kt:40)
        at com.example.login.ui.fragments.GameViewModel.computerMakeTurn(GameViewModel.kt:54)
        at com.example.login.ui.fragments.GameViewModel.onCellClick(GameViewModel.kt:74)
        at com.example.login.ui.fragments.GameFragment.onViewCreated$lambda-2$lambda-1(GameFragment.kt:32)
        at com.example.login.ui.fragments.GameFragment.$r8$lambda$XbcRWBqeqmjzlbSxfrA35PR6S2I(Unknown Source:0)
        at com.example.login.ui.fragments.GameFragment$$ExternalSyntheticLambda1.onClick(Unknown Source:4)
        at android.view.View.performClick(View.java:7455)
        at android.view.View.performClickInternal(View.java:7432)

File: .idea/deploymentTargetDropDown.xml (hunk 1 above)
Comment (podolyanbogdan) [reply .idea/deploymentTargetDropDown.xml:1]: done

File: app/build.gradle (hunk 2 above)
Comment (podolyanbogdan) [reply app/build.gradle:36]: done +-

File: app/src/main/java/com/example/login/canvas/TicTacBoard.kt (hunk 3 above)
Comment (podolyanbogdan) [reply app/src/main/java/com/example/login/canvas/TicTacBoard.kt:50]: done +-

File: app/src/main/java/com/example/login/canvas/TicTacBoard.kt (hunk 4 above)
Comment (podolyanbogdan) [reply app/src/main/java/com/example/login/canvas/TicTacBoard.kt:59]: not done

File: app/src/main/java/com/example/login/ui/fragments/GameFragment.kt (hunk 5 above)
Comment (podolyanbogdan) [reply app/src/main/java/com/example/login/ui/fragments/GameFragment.kt:32]: not done

File: app/src/main/java/com/example/login/ui/fragments/GameViewModel.kt (hunk 6 above)
Comment (podolyanbogdan) [reply app/src/main/java/com/example/login/ui/fragments/GameViewModel.kt:31]: done

File: app/src/main/res/layout/fragment_game.xml (hunk 7 above)
Comment (podolyanbogdan) [reply app/src/main/res/layout/fragment_game.xml:15]: done

**Pull request merged**